### DISCUSSION TITLE
Building a smaller version of Pyodide

### DISCUSSION THREAD

**Discussion by davidfstr** [pyodide/pyodide#5133] [Q&A]
The latest version of Pyodide (0.26.3) looks like it weighs in at about 5 MB compressed† when initially loaded. I'm interested in reducing the initial size down to about 1.5 MB to keep load times down on slow/rural school networks. **Any tips for altering Pyodide's build system to remove/defer parts of Python or its standard library?**

Some research I've done so far:
* I see that certain stdlib modules are already loaded on demand, presumably because of size: openssl, hashlib, sqlite3, pydoc, lzma. Perhaps there's a place in the build system that can select stdlib modules to exclude?
   * Are these modules disabled in: pyodide/cpython/Setup.local ? Or elsewhere?
   * Are these modules added back as packages in files like: pyodide/packages/liblzma/meta.yaml ? Or elsewhere?

Other ideas for stuff to remove:
* Networking. Crypto. I don't need any stdlib modules related to networking.
* Unicode support, except for UTF-8 and perhaps the Windows-1252 encoding.
* Various other Python modules with either a large .py file (inspect.py, typing.py, tarfile.py) or a large C implementation (_datetime, ... probably others)

† Statistics: Sizes obtained from Network panel in Chrome when visiting https://pyodide.org/en/stable/console.html . Biggest files are pyodide.asm.wasm (2.8 MB) and python_stdlib.zip (2.3 MB).

**Comment by hoodmane** [comment]
The biggest module is the encodings module. If you're interested in reducing the size of the standard library you might start by getting rid of all unnecessary encodings. It would probably also help to make a build with no dynamic linking, if you don't need C extension modules.

**Comment by ryanking13** [comment]
> Are these modules disabled in: pyodide/cpython/Setup.local ?

Yes, you can remove modules that you want to exclude there.

> Are these modules added back as packages in files like: pyodide/packages/liblzma/meta.yaml

Yes. [recipes with `cpython_module` type is standard libraries](https://github.com/search?q=repo%3Apyodide%2Fpyodide+cpython_module+language%3AYAML&type=code&l=YAML).

**Comment by davidfstr** [comment]
If you use a disk space visualizer on the contents of python_stdlib.zip , you can see the largest files:

<details><summary>Details</summary>

![Screenshot 2024-10-28 at 1 00 04 PM](https://github.com/user-attachments/assets/7f27b3df-09f8-408f-a832-a8d2973e9160)

</details>

For example, the top 5 seem to be:
1. encodings
2. asyncio
3. email
4. xml
5. multiprocessing

It appears that items can be omitted from python_stdlib.zip using PYZIP_EXCLUDE_FILES.

**Comment by davidfstr** [comment]
Thanks everyone. These are enough tips to get started. I'll close this discussion for now.

**Reply by hoodmane** [reply]
Don't hesitate to open more discussions or issues if you have more questions.
